cpp/src/arrow/compute/exec/options.h

```diff
@@ -199,21 +199,32 @@ class ARROW_EXPORT ProjectNodeOptions : public ExecNodeOptions {
   std::vector<std::string> names;
 };
 
-/// \brief Make a node which aggregates input batches, optionally grouped by keys.
+/// \brief Make a node which aggregates input batches, optionally grouped by keys and
+/// optionally segmented by segment-keys. Both keys and segment-keys determine the group.
+/// However segment-keys are also used for determining grouping segments, which should be
+/// large, and allow streaming a partial aggregation result after processing each segment.
+/// One common use-case for segment-keys is ordered aggregation, in which the segment-key
+/// attribute specifies a column with non-decreasing values or a lexigographically-ordered
+/// set of such columns.
 ///
 /// If the keys attribute is a non-empty vector, then each aggregate in `aggregates` is
 /// expected to be a HashAggregate function. If the keys attribute is an empty vector,
 /// then each aggregate is assumed to be a ScalarAggregate function.
 class ARROW_EXPORT AggregateNodeOptions : public ExecNodeOptions {
  public:
   explicit AggregateNodeOptions(std::vector<Aggregate> aggregates,
-                                std::vector<FieldRef> keys = {})
-      : aggregates(std::move(aggregates)), keys(std::move(keys)) {}
+                                std::vector<FieldRef> keys = {},
+                                std::vector<FieldRef> segment_keys = {})
+      : aggregates(std::move(aggregates)),
+        keys(std::move(keys)),
+        segment_keys(std::move(segment_keys)) {}
 
   // aggregations which will be applied to the targetted fields
   std::vector<Aggregate> aggregates;
   // keys by which aggregations will be grouped
   std::vector<FieldRef> keys;
+  // keys by which aggregations will be segmented
+  std::vector<FieldRef> segment_keys;
```

Comment:
I think referencing the same field should just be disallowed (this is not done) in the PR currently); doing so is not necessary because the output of segmented aggregation with keys `K` and segment-keys `S` is, by definition, equal to that of regular aggregation with keys `K` union `S`.

Reply:
Maybe we should have keys to just be `K` since that matches the original semantics? I think that's where the current confusion comes from.

Reply:
When I see the word "keys" I would natural think it was the grouping key, e.g., `K`, instead of "keys that is not segmented key", i.e, `K - S`

Reply:
Ok.  So if I want to group by A, B, C and I know my data is sorted by A then I can do segments=A and keys=B C?

In that case I agree we should explicitly disallow this to help future users.  Also, we might want to improve the description of these fields a bit and provide some examples, but that isn't urgent for right now.

Reply:
I added the check to disallow.

Reply:
Can we document this in AggregateNodeOptions and make it super clear?